pretyping/pretyping.ml

```diff
@@ -870,6 +902,20 @@ struct
           let sigma, hj = pretype empty_tycon env sigma c in
           error_cant_apply_not_functional
             ?loc:(Loc.merge_opt floc argloc) !!env sigma resj [|hj|]
+        in
+        match impls with
+        | None -> f c rest
+        | Some (coe,_) ->
+          let c',rest',some_change = insert_impargs ?loc impls c rest in
+          if not some_change then f c rest else
+          let g c rest = try Inl (f c rest) with e when CErrors.noncritical e -> Inr e in
```

Comment:
Please avoid generic catch-all of exceptions.

Reply:
Sorry, I did not get it. What is your suggestion?

Reply:
As @ppedrot suggested before `f` should not raise and instead return an algebraic type; the code indeed looks pretty hard to reason as calls to `user_err` are later absorbed.

Reply:
Moreover the reraise is likely incorrect below, as in it loses information. That should not matter for Coq, but unfortunately that may not be the case.

Reply:
> As @ppedrot suggested before f should not raise and instead return an algebraic type

I don't see how I can avoid `f` to raise an exception. `f` is calling `unify`, `pretype`, all kinds of function which possibly raise an exception w/o having control on where it is raised.

> as in it loses information

You mean information of the trace?

By the way, I meant reraising `e'` not `e`. Changing.

Reply:
> I don't see how I can avoid `f` to raise an exception. `f` is calling `unify`, `pretype`, all kinds of function which possibly raise an exception w/o having control on where it is raised.

Do we have an specification of what these functions can raise? Indeed they should not raise except on error, and if there is an error there why we absorb it? IMHO we really have a mess in the codebase. I cannot understand for example why here all errors are actually absorbed, that seems wrong in a few cases, like for example a missing lib by `Coqlib` etc... Yeah our notion of "recoverable error" is pretty ill-defined.

I know that is it not easy to fix this, but at least IMO in new code the `try ... with ...` block should be pushed to the inner boundaries so they act as wrappers of the raising functions and turn the caught exception into something sensible.

> You mean information of the trace?

Of the trace and anything else users attached to it using our dangerous Exnifo API.

Reply:
> Do we have an specification of what these functions can raise?

You know how helpful is OCaml to inform about side effects!

In any case, this could be a catch of `precatchable_exception` instead of `noncritical` but I'd bet that this would not make a difference.

Are you suggesting that we should pass not only states but exceptions in purely functional style?

Reply:
>if there is an error there why we absorb it

It's done that way to implement the compatibility mode. 
IMO this is the wrong way to work, if we want the PR's feature we should put an option to disable it and deprecate the option. Then we don't need to catch exceptions here.

Reply:
Having a warning with a location is useful to upgrade. Then, we could do an automatic translation, the same way the python script replaced automatically references to deprecated aliases in 8.9. An option is possible in parallel.